### PR TITLE
group_by_rows(...).aggregate(...) in the IR

### PR DESCRIPTION
**Pull request by danking**
Created a Relational IR for `mt.group_by_rows(...).aggregate(...)`.

I've never done this before so I just copied and massaged `MatrixMapRows`. It seems a lot longer than I anticipated. Two of the `GroupBySuite` tests exercise this path. The other two use non-IR features. When more things are put into the expr IR, they'll start exercising this path too.

### CODE REVIEW

File: src/main/scala/is/hail/expr/Relational.scala [hunk 1]
```diff
@@ -626,6 +627,213 @@ case class CollectColsByKey(child: MatrixIR) extends MatrixIR {
   }
 }
 
+case class MatrixAggregateRowsByKey(child: MatrixIR, expr: IR) extends MatrixIR {
+  def children: IndexedSeq[BaseIR] = Array(child, expr)
+
+  def copy(newChildren: IndexedSeq[BaseIR]): MatrixAggregateRowsByKey = newChildren match {
+    case Seq(child: MatrixIR, aggExpr: IR) =>
```
[the remark this answers is not in the document]
Comment (rcownie) [reply src/main/scala/is/hail/expr/Relational.scala]: OK, I'm more caffeinated today and I see there's precisely one child and what is matched by aggExpr is the original expr.

**Comment by rcownie** [comment]
This all looks ok to me except for the two "global" params.  I think if you fix that both in the Compile
and the call, and it passes tests, then we're good to go.

**Comment by rcownie** [comment]
I hit the wrong button "close and comment".  Should stay open.

**Comment by danking** [comment]
I also am not super familiar with this code, but I think
```scala
   val (rvAggs, seqOps, aggResultType, f, rTyp) = ir.CompileWithAggregators[Long, Long, Long, Long, Long, Long](
      "global", child.typ.globalType,
      "AGG", tAgg,
      "global", child.typ.globalType,
      "colValues", TArray(child.typ.colType),
      "va", child.typ.rvRowType,
```
is actually two separate argument lists. The first is just `"globals"` which is the argument to the snippet of IR that is executed _after_ aggregation is finished. The second four arguments are the arguments to the aggregation IR which needs the aggregable itself, the globals, the fields for every column, and the fields of the row currently-being-aggregated-over.

**Comment by danking** [comment]
(truth be told, I copy-pasted all of this from `MatrixMapRows`)

**Comment by danking** [comment]
@rcownie I am now feeling positive about this change. If you're still happy with it lets approve and merge.

**Comment by danking** [comment]
For posterity, I ran this test:
```python
ds = hl.read_matrix_table('gs://danking/ALL.chip.omni_broad_sanger_combined.20140818.snps.genotypes-hail-bff300d475ac.mt')
gp = (ds.group_rows_by(contig=ds.locus.contig, pos=ds.locus.position/1000)
        .aggregate(n_non_ref = hl.agg.count_where(ds.GT.is_non_ref())))
gp = gp.filter_entries((gp.n_non_ref != 0) & hl.is_defined(gp.n_non_ref))
gp.n_non_ref.show()
```
```bash
(hail) 1 dking@wmb16-359 # gsutil du -sh gs://danking/ALL.chip.omni_broad_sanger_combined.20140818.snps.genotypes-hail-bff300d475ac.mt
1.23 GiB    gs://danking/ALL.chip.omni_broad_sanger_combined.20140818.snps.genotypes-hail-bff300d475ac.mt
```

**Comment by danking** [comment]
ping @rcownie

File: src/main/scala/is/hail/expr/Relational.scala [hunk 2]
```diff
@@ -708,6 +709,239 @@ case class CollectColsByKey(child: MatrixIR) extends MatrixIR {
...
+
+  private[this] val tAggElt: Type = child.typ.entryType
+  // FIXME: this should come from the MatrixType
+  private[this] val aggSymTab = Map(
+    "global" -> (0, child.typ.globalType),
+    "va" -> (1, child.typ.rvRowType),
+    "g" -> (2, child.typ.entryType),
+    "sa" -> (3, child.typ.colType))
+
+  private[this] val tAgg = TAggregable(tAggElt, aggSymTab)
```
[the remark this answers is not in the document]
Comment (danking) [reply src/main/scala/is/hail/expr/Relational.scala]: Good catch! I rebase failure on my part.

**Pull request merged**